daemon/src/maker_inc_connections.rs

```diff
@@ -102,43 +111,40 @@ impl Actor {
         let transport_state = Arc::new(Mutex::new(transport_state));
 
         let (read, write) = stream.into_split();
-        let read = FramedRead::new(read, wire::EncryptedJsonCodec::new(transport_state.clone()))
-            .map_ok(move |msg| FromTaker { taker_id, msg })
-            .map(forward_only_ok::Message);
+        let mut read =
+            FramedRead::new(read, wire::EncryptedJsonCodec::new(transport_state.clone()))
+                .map_ok(move |msg| FromTaker { taker_id, msg });
+
+        let this = ctx.address().expect("self to be alive");
+        let taker_msg_channel = self.taker_msg_channel.clone_channel();
+        let read_fut = async move {
+            while let Ok(res) = read.try_next().await {
+                match res {
+                    Some(msg) => taker_msg_channel
+                        .send(msg)
+                        .await
+                        .expect("connection to taker to exist"),
+                    None => continue,
+                }
+            }
 
-        let (out_msg_actor_address, mut out_msg_actor_context) = xtra::Context::new(None);
+            let _ = this.send(ReadFail(taker_id)).await;
+        };
+        taker_tasks.add(read_fut);
 
-        let (forward_to_cfd, forward_to_cfd_fut) =
-            forward_only_ok::Actor::new(self.taker_msg_channel.clone_channel())
-                .create(None)
-                .run();
-        self.tasks.add(forward_to_cfd_fut);
+        let (out_msg, mut out_msg_actor_context) = xtra::Context::new(None);
+        let send_to_socket_actor = send_to_socket::Actor::new(write, transport_state.clone());
 
-        // only allow outgoing messages while we are successfully reading incoming ones
         let heartbeat_interval = self.heartbeat_interval;
-        let taker_disconnected_channel = self.taker_disconnected_channel.clone_channel();
-        self.tasks.add(async move {
-            let mut actor = send_to_socket::Actor::new(write, transport_state.clone());
+        let heartbeat_fut = out_msg_actor_context
+            .notify_interval(heartbeat_interval, || wire::MakerToTaker::Heartbeat)
+            .expect("actor not to shutdown");
+        taker_tasks.add(heartbeat_fut);
 
-            let _heartbeat_handle = out_msg_actor_context
-                .notify_interval(heartbeat_interval, || wire::MakerToTaker::Heartbeat)
-                .expect("actor not to shutdown")
-                .spawn_with_handle();
+        taker_tasks.add(out_msg_actor_context.run(send_to_socket_actor));
```

Comment:
Could we re-arrange this so that all `tasks.add` are in the same place? I think that would be more readable.

Reply:
Do we have a constructor on `Tasks` that takes a `Vec`? Could remove the mutability of `tasks` then and construct and store it in the `HashMap` in one go.

Reply:
Something like this would be nice and compiles

```rust
pub fn new(fs: Vec<impl Future<Output = ()> + Send + 'static>) -> Self {
    let handles = fs.into_iter().map(|f| f.spawn_with_handle()).collect();
    Self(handles)
}
```

but in practice

```
distinct uses of `impl Trait` result in different opaque types
```

Any way we can get around that?

Reply:
You'd have to `Box` which is quite inconvenient.

Could do an `trait IntoHandles` that converts to a `Vec<RemoteHandle>` and is implement for `(A: Future)`, `(A: Future, B: Future)`, etc. Then you could pass in tuples of various sizes ...

Reply:
I'm going to pass on this one, but it's an interesting idea.